module/zfs/zvol.c

```diff
@@ -1218,8 +1218,16 @@ zvol_ioctl(struct block_device *bdev, fmode_t mode,
 	rw_enter(&zv->zv_suspend_lock, RW_READER);
 	switch (cmd) {
 	case BLKFLSBUF:
-		zil_commit(zv->zv_zilog, ZVOL_OBJ);
+		fsync_bdev(bdev);
+
+		if (dsl_dataset_is_dirty(dmu_objset_ds(zv->zv_objset)) &&
+		    !(zv->zv_flags & ZVOL_RDONLY))
+			txg_wait_synced(dmu_objset_pool(zv->zv_objset), 0);
+
+		invalidate_bdev(bdev);
```

Comment:
We don't want zv_suspend_lock to cover fsync_bdev, invalidate_bdev and copy_to_user.